### PR TITLE
Eliminate many possible sources of deadlocks

### PR DESCRIPTION
**Pull request by nickstenning**
Simplify the WebSocket server test code by moving all WebSocket interaction onto the main thread.

Despite previous fixes, we were still seeing frequent deadlocks of Smokey's websocket code on Linux. It seems that at least some of these deadlocks may have been caused by incorrect error handling. Specifically, if a connection failed due to networking problems, this error was swallowed, and a subsequent attempt to close the already-closed connection would deadlock.

This issue could be fixed by adding a mechanism to pass exception data back to the main thread, but this approach would add yet more complexity to some code that is already hard to follow. It seems simpler to eliminate all the multi-process and multi-threading elements of this code instead.

This commit moves all the handling of WebSocket communication onto the main thread, using asyncio's `run_until_complete` function to schedule the relevant coroutines at appropriate times. This should ensure that any exceptions thrown within a coroutine are rethrown inside the behave test functions.

In addition, all the coroutines which make network calls are now wrapped in `asyncio.wait_for` with a timeout, which should ensure blocking forever on I/O isn't possible either.

### CODE REVIEW

File: smokey/features/steps/websocket.py [hunk 1]
```diff
@@ -14,52 +12,13 @@
...
+async def websocket_connect(endpoint, timeout=10, extra_headers=None, verify=True):
     """Establish a websocket connection and send a client_id message."""
-    ssl_context = _ssl_context(verify=verify)
-    websocket = await websockets.connect(endpoint,
-                                         extra_headers=extra_headers,
-                                         ssl=ssl_context)
+    kwargs = {'extra_headers': extra_headers}
+    if endpoint.startswith('wss://'):
+        kwargs['ssl'] = _ssl_context(verify=verify)
+    connect_coro = websocket.connect(endpoint, **kwargs)
```
Comment (chdorner) [inline smokey/features/steps/websocket.py]: I think you mean `websockets.connect(...)`? Just testing this out locally

File: smokey/features/steps/websocket.py [hunk 2]
```diff
@@ -157,15 +104,19 @@ def wait_for_notification(context, delay):
...
+        timeout = deadline - time.monotonic()
+        if timeout <= 0:  # past the deadline
+            break
         try:
-            msg = context.websocket_messages.get(timeout=0.1)
-        except queue.Empty:
-            if time.time() > timeout:
-                break
+            msg_coro = websocket_await_message(context.websocket, timeout)
+            msg = loop.run_until_complete(msg)
```
Comment (chdorner) [inline smokey/features/steps/websocket.py]: And I think here you want to do `loop.run_until_complete(msg_coro)`?

File: smokey/features/steps/websocket.py (hunk 1 above)
Comment (nickstenning) [inline smokey/features/steps/websocket.py]: Jeez. Yes.

File: smokey/features/steps/websocket.py (hunk 2 above)
Comment (nickstenning) [inline smokey/features/steps/websocket.py]: Also 🤦‍♂️.
